File: aos/events/ping.cc
#include "absl/flags/flag.h"

#include "aos/configuration.h"
#include "aos/events/ping_lib.h"
#include "aos/events/shm_event_loop.h"
#include "aos/init.h"

ABSL_FLAG(std::string, config, "pingpong_config.json", "Path to the config.");

int main(int argc, char **argv) {
  aos::InitGoogle(&argc, &argv);
  aos::EventLoop::SetDefaultVersionString("ping_version");

  aos::FlatbufferDetachedBuffer<aos::Configuration> config =
      aos::configuration::ReadConfig(absl::GetFlag(FLAGS_config));

  aos::ShmEventLoop event_loop(&config.message());

  aos::Ping ping(&event_loop);

  event_loop.Run();

  return 0;
}

File: aos/events/ping.rs
use aos::configuration;
use aos::events::shm_event_loop::ShmEventLoop;
use aos::init::Init;
use clap::Parser;
use ping_lib::PingTask;
use std::path::Path;

/// Ping portion of a ping/pong system.
#[derive(Parser, Debug)]
struct App {
    /// Time to sleep between pings.
    #[arg(long, default_value_t = 10000, value_name = "MICROS")]
    sleep: u64,
}

fn main() {
    let app = App::init();
    let config = configuration::read_config_from(Path::new("pingpong_config.json")).unwrap();
    let ping = PingTask::new();
    ShmEventLoop::new(&config).run_with(|runtime| {
        runtime.set_realtime_priority(5);
        runtime.spawn(ping.tasks(*runtime, app.sleep));
    });
}
